usr/lib/cca_stdll/cca_specific.c

```diff
@@ -3988,331 +4607,553 @@ static CK_RV build_public_EC_key_value_structure(CK_BYTE *pubkey, CK_ULONG puble
-                              &opaque_key))) {
-        TRACE_DEVEL("build_attribute(CKA_IBM_OPAQUE) failed\n");
-        return rc;
-    }
-    rc = template_update_attribute(pub_templ, opaque_key);
-    if (rc != CKR_OK) {
-        TRACE_DEVEL("template_update_attribute(CKA_IBM_OPAQUE) failed\n");
-        free(opaque_key);
-        return rc;
+        rc = build_public_EC_key_value_structure(pubkey, publen,
+                                                 curve_type, curve_bitlen,
+                                                 (unsigned char *)&key_value_structure,
+                                                 &key_value_structure_length);
+        if (rc != CKR_OK)
+            return rc;
+
+        /* Build public key token */
+        rule_array_count = 1;
+        memcpy(rule_array, "ECC-PUBL", (size_t)(CCA_KEYWORD_SIZE));
+        private_key_name_length = 0;
+        key_token_length = CCA_KEY_TOKEN_SIZE;
+        key_value_structure_length = CCA_KEY_VALUE_STRUCT_SIZE;
+
+        dll_CSNDPKB(&return_code, &reason_code,
+                    &exit_data_len, exit_data,
+                    &rule_array_count, rule_array,
+                    &key_value_structure_length, key_value_structure,
+                    &private_key_name_length, private_key_name,
+                    &param1, param2, &param1, param2, &param1, param2,
+                    &param1, param2, &param1, param2,
+                    &key_token_length,
+                    key_token);
+
+        if (return_code != CCA_SUCCESS) {
+            TRACE_ERROR("CSNDPKB (EC KEY TOKEN BUILD) failed. return:%ld,"
+                        " reason:%ld\n", return_code, reason_code);
+            if (is_curve_error(return_code, reason_code))
+                return CKR_CURVE_NOT_SUPPORTED;
+            return CKR_FUNCTION_FAILED;
+        }
+
+        /* Public keys do not need to be wrapped, so just add public
+           key token to template as CKA_IBM_OPAQUE */
+        if ((rc = build_update_attribute(pub_templ, CKA_IBM_OPAQUE,
+                                         key_token, key_token_length))) {
+            TRACE_DEVEL("build_update_attribute(CKA_IBM_OPAQUE) failed\n");
+            return rc;
+        }
     }
 
+#ifdef DEBUG
+    TRACE_DEBUG("%s: imported object template attributes:\n", __func__);
+    dump_template(pub_templ);
+#endif
+
     return CKR_OK;
 }
 
 CK_RV token_specific_object_add(STDLL_TokData_t *tokdata, SESSION *sess, OBJECT *object)
 {
-	CK_RV rc;
-	CK_KEY_TYPE keytype;
-	CK_OBJECT_CLASS keyclass;
+    CK_RV rc;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_KEY_TYPE keytype;
+    CK_OBJECT_CLASS keyclass;
 
-	UNUSED(tokdata);
-	UNUSED(sess);
+    UNUSED(tokdata);
+    UNUSED(sess);
 
-	if (!object) {
-		TRACE_ERROR("Invalid argument\n");
-		return CKR_FUNCTION_FAILED;
-	}
+    if (!object) {
+        TRACE_ERROR("Invalid argument\n");
+        return CKR_FUNCTION_FAILED;
+    }
 
+    /* we only deal with key objects here */
```

Comment:
Actually, you could check here already if an CKA_IBM_OPAQUE is there, and if so, then analyze it and find out about the key type and class from the token. This would lift the requirement for the user to specify CLA_KEY_TYPE and CKA_CLASS when importing an opaque key. The key token does in fact contain all this information already, so no need for the user to specify it.
The only question would be if an EP11 key token also allows to get the same kind of information from it. So once we do the key blob import for EP11, it then should behave the same as for CCA.

Reply:
I'll leave this to Reinhard to decide. It's doable but then a user does not need and provide any key information but only the CCA secure key token. So no way to check if this was the key type the user intended to import.